docs/RFCS/20220628_invisible_index.md

```diff
@@ -0,0 +1,468 @@
+  - force index with invisible index
+  - having a foreign key constraint referencing a unique invisible index
+    - having a unique invisible index in general (warn if you create a child table referencing it, warn if you are using insert on conflict with it, delete on cascade)
+
+### 1. What types of indexes can be invisible?
+In [MySQL](https://dev.mysql.com/doc/refman/8.0/en/invisible-indexes.html), all
+indexes other than primary keys explicitly or implicitly can be invisible. In
+MySQL, a table with no explicit primary key will use the first unique index on
+NOT NULL columns as an implicit primary key. Implicit and explicit primary
+indexes are both not allowed to be invisible.
+
+In CockroachDB, a table with no explicit primary key creates a new rowid column,
+and creating a unique index on a null column will not change the primary key.
+Primary indexes cannot be invisible. Unique indexes can be invisible.
+
+### 2. Are constraints allowed to be created with invisible indexes?
+In MySQL, only indexes can be invisible.
+
+In CockroachDB, only indexes can be invisible as well. Creating a unique
+constraint or a foreign key constraint with invisible indexes does not make
+sense and will not be supported.
+
+One might think creating unique constraints with invisible indexes is similar to
+creating unique constraints without indexes (which is something we are currently
+supporting). But they have very different semantic meanings. Firstly, creating a
+constraint without index was not user-friendly and was created for multi-tenant
+testing purposes. Secondly, creating a constraint with an invisible index is
+still an index but just ignored by the optimizer.
+
+This leads to an issue with the parser. Creating an invisible unique index
+inside a `CREATE TABLE` definition is supported by the grammar rule, but the
+parser will throw an error. This is because the parser is doing a round trip in
+`pkg/testutils/sqlutils/pretty.go`. In `sql.y`, creating an unique index in a
+`CREATE TABLE` statement returns a new structure
+`tree.UniqueConstraintTableDef`. However, creating a unique constraint with a
+not visible index is not supported by the parser. When the parser does a round
+trip for the following statement, it formats it to a pretty statement using the
+unique constraint definition. But the parser does not support unique constraints
+with not visible index syntax. So it will fail while parsing the pretty
+statement. Since `logictest` also performs a roundtrip check in
+`pkg/sql/logictest/logic.go`, `logictest` would also fail. But creating an
+unique index inside a `CREATE TABLE` definition will still work in a cluster.
+This is a known issue. See more details in
+https://github.com/cockroachdb/cockroach/pull/65825. The current solution for
+this issue is to not create any tests in parser or in logictest. We will test
+invisible unique indexes with `CREATE UNIQUE INDEX` and hope the behavior will
```

Comment:
We should have some tests for `CREATE TABLE` with unique invisible indexes, rather than just hoping it will work. Can we add an option to logic tests to disable the round-tripping for a given statement, like `statement ok disable-round-trip`.